File: lib/aihelper.rb
require 'json'
require_relative 'openai'
require 'rest-client'

OpenAI.configure do |config|
  config.access_token = 'OPENAI_KEY_REDACTED'
  config.organization_id = 'org-AQP4h0lMfOWuCW9WabhnxRRl'
end

client = OpenAI::Client.new

# puts client.models.list['data'].map{|data| data['id'] }

# client.chat(
#     parameters: {
#         model: "gpt-3.5-turbo", # Required.
#         messages: [{ role: "user", content: "Напиши хоку"}], # Required.
#         temperature: 0.7,
#         stream: proc do |chunk, _bytesize|
#             print chunk.dig("choices", 0, "delta", "content")
#         end
#     })

client.chat(
    parameters: {
        model: "gpt-4", # Required.
        messages: [{ role: "user", content: <<-EOF

        
        EOF
        }], # Required.
        temperature: 0.9,
        stream: proc do |chunk, _bytesize|
            print chunk.dig("choices", 0, "delta", "content")
        end
    })

# response = client.images.generate(parameters: { prompt: "Сделай иллюстрацию к новости: Официальный представитель МИД РФ Мария Захарова заявила в своем Telegram-канале, что в списках на эвакуацию из сектора Газа числится очень много детей.
# Как передавало ИА Регнум, Египет 1 ноября открыл терминал КПП «Рафах» на границе с сектором Газа, чтобы впустить около 90 раненых палестинцев и около 450 иностранцев.
# Девятого ноября Захарова сказала, что в МИД РФ шокированы заявлением посла Израиля о сроках эвакуации россиян из Газы.
# 8 ноября посол Израиля в Москве Александр Бен Цви заявил, что согласование эвакуации граждан РФ из сектора Газа может занять до двух недель.", size: "1024x1024" })
# puts response.dig("data", 0, "url")
# => "https://oaidalleapiprodscus.blob.core.windows.net/private/org-Rf437IxKhh..."

# OpenAI.rough_token_count("Your text")

# response = client.chat(
#     parameters: {
#         model: "gpt-3.5-turbo", # Required.
#         messages: [{ role: "user", content: "Say this is a test!"}], # Required.
#         temperature: 0.7,
#     })
# puts response.dig("choices", 0, "message", "content")

# response = client.audio.translate(
#     parameters: {
#         model: "whisper-1",
#         file: File.open("./tiroler.m4a", "rb"),
#         # language: 'ru',
#     })
# puts response["text"]

# response = client.audio.transcribe(
#     parameters: {
#         model: "whisper-1",
#         file: File.open("./tiroler.m4a", "rb"),
#     })
# puts response["text"]

# response = client.speech.create(
#     parameters: {
#         model: "tts-1",
#         input: 'Сегодня КЛАССНЫЙ день!, малыш я счастлива что ТЫ у меня есть!', # "Today is a wonderful day to build something people love!",
#         voice: "nova" # alloy, echo, fable, onyx, nova, and shimmer
#     })
# File.open('output.mp3', 'w') { |file| file.write(response) }


# client.chat(
#   parameters: {
#     model: "gpt-4-vision-preview", # Required.
#     messages: [{ role: "user", content: [
#       {
#         "type": "text",
#         "text": "What are in these images? Is there any difference between them?"
#       },
#       {
#         "type": "image_url",
#         "image_url": {
#           "url": "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg",
#         }
#       },
#       {
#         "type": "image_url",
#         "image_url": {
#           "url": "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg",
#         }
#       }
#     ]}],
#     temperature: 0.7,
#     stream: proc do |chunk, _bytesize|
#       print chunk.dig("choices", 0, "delta", "content")
#     end
#   }
# )


# # Example dummy function hard coded to return the same weather
# # In production, this could be your backend API or an external API
# def get_current_weather(location:, unit: "fahrenheit")
#   if location.downcase.include? "tokyo"
#     return {"location" => "Tokyo", "temperature" => "10", "unit" => "celsius"}
#   elsif location.downcase.include? "san francisco"
#     return {"location" => "San Francisco", "temperature" => "72", "unit" => "fahrenheit"}
#   elsif location.downcase.include? "paris"
#     return {"location" => "Paris", "temperature" => "22", "unit" => "celsius"}
#   else
#     return {"location" => location, "temperature" => "unknown"}
#   end
# end

# def run_conversation()
#   # Step 1: send the conversation and available functions to the model
#   messages = [{"role" => "user", "content" => "What's the weather like in San Francisco, Tokyo, and Paris?"}]
#   tools = [
#       {
#           "type" => "function",
#           "function" => {
#               "name" => "get_current_weather",
#               "description" => "Get the current weather in a given location",
#               "parameters" => {
#                   "type" => "object",
#                   "properties" => {
#                       "location" => {
#                           "type" => "string",
#                           "description" => "The city and state, e.g. San Francisco, CA",
#                       },
#                       "unit" => {"type" => "string", "enum" => ["celsius", "fahrenheit"]},
#                   },
#                   "required" => ["location"],
#               },
#           },
#       }
#   ]

#   client = OpenAI::Client.new
#   response = client.completion.create(parameters: {
#     model: "gpt-3.5-turbo-1106",
#     messages: messages,
#     tools: tools,
#     tool_choice: "auto"  # auto is default, but we'll be explicit
#   })
#   response_message = response.dig("choices", 0, "message")
#   tool_calls = response.dig("choices", 0, "message", "tool_calls")

#   # Step 2: check if the model wanted to call a function
#   if tool_calls
#     # Step 3: call the function
#     # Note: the JSON response may not always be valid; be sure to handle errors
#     available_functions = {
#         "get_current_weather" => method(:get_current_weather),
#     }
#     messages.append(response_message) # extend conversation with assistant's reply

#     # Step 4: send the info for each function call and function response to the model
#     tool_calls.each do |tool_call|
#       function_name = tool_call.dig("function", "name")
#       function_to_call = available_functions[function_name]
#       function_args = JSON.parse(tool_call.dig("function", "arguments"))
#       function_response = function_to_call.call(
#           location: function_args["location"],
#           unit: function_args["unit"],
#       )
#       messages.append(
#           {
#               "tool_call_id" => tool_call['id'],
#               "role" => "tool",
#               "name" => function_name,
#               "content" => function_response.to_json,
#           }
#       )
#     end

#     second_response = client.completion.create(parameters: {
#         model: "gpt-3.5-turbo-1106",
#         messages: messages,
#     })
#     return second_response
#   end
# end

# response = run_conversation()

# puts response.dig("choices", 0, "message", 'content') # => The current weather in San Francisco is 72°F, in Tokyo it is 10°C, and in Paris it is 22°C.



# сделать ассистента который выполняет код на python

File: lib/aihelper.py
from openai import OpenAI
from dotenv import load_dotenv

load_dotenv()

# import os
# print(os.getenv("OPENAI_API_KEY"))  # should print your API key

client = OpenAI(
  # Defaults to os.environ.get("OPENAI_API_KEY")
  # Otherwise use: api_key="Your_API_Key",
)

# response = client.chat.completions.create(
#   model="gpt-3.5-turbo",
#   messages=[
#         {"role": "user", "content": "Напиши хоку"},
#     ],
#   temperature=0.7
# )
# print(response.choices[0].message.content)

##############################################

# import json

# # Example dummy function hard coded to return the same weather
# # In production, this could be your backend API or an external API
# def get_current_weather(location, unit="fahrenheit"):
#     """Get the current weather in a given location"""
#     if "tokyo" in location.lower():
#         return json.dumps({"location": location, "temperature": "10", "unit": "celsius"})
#     elif "san francisco" in location.lower():
#         return json.dumps({"location": location, "temperature": "72", "unit": "fahrenheit"})
#     else:
#         return json.dumps({"location": location, "temperature": "22", "unit": "celsius"})

# def run_conversation():
#   # Step 1: send the conversation and available functions to the model
#   messages = [{"role": "user", "content": "What's the weather like in San Francisco, Tokyo, and Paris?"}]
#   tools = [
#       {
#           "type": "function",
#           "function": {
#               "name": "get_current_weather",
#               "description": "Get the current weather in a given location",
#               "parameters": {
#                   "type": "object",
#                   "properties": {
#                       "location": {
#                           "type": "string",
#                           "description": "The city and state, e.g. San Francisco, CA",
#                       },
#                       "unit": {"type": "string", "enum": ["celsius", "fahrenheit"]},
#                   },
#                   "required": ["location"],
#               },
#           },
#       }
#   ]
#   response = client.chat.completions.create(
#       model="gpt-4-1106-preview", # gpt-4-1106-preview
#       messages=messages,
#       tools=tools,
#       tool_choice="auto",  # auto is default, but we'll be explicit
#   )
#   response_message = response.choices[0].message
#   tool_calls = response_message.tool_calls
#   # Step 2: check if the model wanted to call a function
#   if tool_calls:
#       # Step 3: call the function
#       # Note: the JSON response may not always be valid; be sure to handle errors
#       available_functions = {
#           "get_current_weather": get_current_weather,
#       }  # only one function in this example, but you can have multiple
#       messages.append(response_message)  # extend conversation with assistant's reply
#       # Step 4: send the info for each function call and function response to the model
#       for tool_call in tool_calls:
#           function_name = tool_call.function.name
#           function_to_call = available_functions[function_name]
#           function_args = json.loads(tool_call.function.arguments)
#           function_response = function_to_call(
#               location=function_args.get("location"),
#               unit=function_args.get("unit"),
#           )
#           messages.append(
#               {
#                   "tool_call_id": tool_call.id,
#                   "role": "tool",
#                   "name": function_name,
#                   "content": function_response,
#               }
#           )  # extend conversation with function response
#       second_response = client.chat.completions.create(
#           model="gpt-4-1106-preview", # gpt-4-1106-preview
#           messages=messages,
#       )  # get a new response from the model where it can see the function response
#       return second_response

# response = run_conversation()
# print(response.choices[0].message.content)

##############################

# messages = [
#   {"role": "system", "content": "Ты полезный помощник"},
#   {"role": "user", "content": "Кто выиграл мировую серию в 2020?"},
#   {"role": "assistant", "content": "The Los Angeles Dodgers won the World Series in 2020."},
#   {"role": "user", "content": "Где играли?"}
# ]

# response = client.chat.completions.create(
#   model = "gpt-3.5-turbo-16k-0613",
#   messages = messages
# )

# print(response.choices[0].message.content)
# messages.append(response.choices[0].message)
# messages.append({"role": "user", "content": "Сколько набрали очков?"})
# print('Сколько набрали очков?')

# response = client.chat.completions.create(
#   model = "gpt-3.5-turbo-16k-0613",
#   messages = messages
# )

# print(response.choices[0].message.content)

# ############################################

# messages = [
#   {"role": "system", "content": "Ты полезный помощник"},
#   {"role": "user", "content": "Кто выиграл мировую серию в 2020?"},
#   {"role": "assistant", "content": "The Los Angeles Dodgers won the World Series in 2020."},
#   {"role": "user", "content": "Где играли?"}
# ]

# response = client.chat.completions.create(
#   model = "gpt-3.5-turbo-16k-0613",
#   messages = messages
#   stream = True
# )
# for chunk in response:
#   print(chunk.choices[0].delta.content, end = "")


############################## Assistant

assistant = client.beta.assistants.create(
  name="Math Tutor",
  instructions="You are a personal math tutor. Write and run code to answer math questions.",
  tools=[{"type": "code_interpreter"}],
  model="gpt-4-1106-preview"
)

thread = client.beta.threads.create()

message = client.beta.threads.messages.create(
  thread_id=thread.id,
  role="user",
  content="I need to solve the equation `3x + 11 = 14`. Can you help me?"
)

run = client.beta.threads.runs.create(
  thread_id=thread.id,
  assistant_id=assistant.id,
  instructions="Please address the user as Jane Doe. The user has a premium account."
)

run = client.beta.threads.runs.retrieve(
  thread_id=thread.id,
  run_id=run.id
)

# wait until run status='completed'

messages = client.beta.threads.messages.list(
  thread_id=thread.id
)

print(messages)
